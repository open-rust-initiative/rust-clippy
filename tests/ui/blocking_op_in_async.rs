#![feature(async_fn_in_trait)]
#![feature(async_closure)]
#![warn(clippy::blocking_op_in_async)]
#![allow(clippy::let_underscore_future)]
#![allow(incomplete_features)]

use std::fs::read;
use std::thread::sleep;
use std::time::Duration;

pub async fn async_std_sleep() {
    sleep(Duration::from_secs(1));
}

mod totally_thread_safe {
    use std::time::Duration;

    pub async fn sleep(_dur: Duration) {}
}

mod custom_tests {
    use super::totally_thread_safe::sleep;
    use std::time::Duration;

    pub async fn conflicted_name() {
        sleep(Duration::from_secs(1)); // don't lint this
    }
}

trait AsyncTrait {
    async fn foo(&self);
    async fn bar(&mut self);
}

struct SomeType(u8);

impl AsyncTrait for SomeType {
    async fn foo(&self) {
        sleep(Duration::from_secs(self.0 as _));
    }
    // don't lint
    async fn bar(&mut self) {
        self.0 = 1;
    }
}

fn do_something() {}

#[rustfmt::skip]
fn closures() {
    let _ = async || {
        sleep(Duration::from_secs(1))
    };
    let async_closure = async move |_a: i32| {
        let _ = 1;
        do_something();
        sleep(Duration::from_secs(1));
    };
    // don't lint, not async
    let non_async_closure = |_a: i32| {
        sleep(Duration::from_secs(1));
        do_something();
    };
}

fn main() {}
