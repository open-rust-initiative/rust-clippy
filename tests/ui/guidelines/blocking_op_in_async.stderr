error: this function call might blocks the thread in async context
  --> $DIR/blocking_op_in_async.rs:12:5
   |
LL |     sleep(Duration::from_secs(1));
   |     ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: asyncness was determined here
  --> $DIR/blocking_op_in_async.rs:11:1
   |
LL | pub async fn async_std_sleep() {
   | ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   = note: `-D clippy::blocking-op-in-async` implied by `-D warnings`

error: this function call might blocks the thread in async context
  --> $DIR/blocking_op_in_async.rs:39:9
   |
LL |         sleep(Duration::from_secs(self.0 as _));
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: asyncness was determined here
  --> $DIR/blocking_op_in_async.rs:38:5
   |
LL |     async fn foo(&self) {
   |     ^^^^^^^^^^^^^^^^^^^

error: this function call might blocks the thread in async context
  --> $DIR/blocking_op_in_async.rs:52:9
   |
LL |         sleep(Duration::from_secs(1))
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: asyncness was determined here
  --> $DIR/blocking_op_in_async.rs:51:13
   |
LL |     let _ = async || {
   |             ^^^^^

error: this function call might blocks the thread in async context
  --> $DIR/blocking_op_in_async.rs:57:9
   |
LL |         sleep(Duration::from_secs(1));
   |         ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
   |
note: asyncness was determined here
  --> $DIR/blocking_op_in_async.rs:54:25
   |
LL |     let async_closure = async move |_a: i32| {
   |                         ^^^^^

error: aborting due to 4 previous errors

